### PR TITLE
Reduce backscatter by checking return-path domain SPF status

### PR DESCRIPTION
**Pull request by acasajus**
Mute 5XX return codes with 2XX codes when return-path domain fails SPF checks

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -257,6 +257,44 @@ def get_string_dict():
...
+            "R_SPF_NA": SPFCheckResult.not_available,
+            "R_SPF_PERMFAIL": SPFCheckResult.perm_error,
+        }
+
+
+class SpamdResult:
+    def __init__(self):
+        self.dmarc: DmarcCheckResult = DmarcCheckResult.not_available
+        self.spf: SPFCheckResult = SPFCheckResult.not_available
+        self.domain = "unknown"
```
Comment (nguyenkims) [inline app/models.py:286]: We don't set the domain yet right? I wonder if we should distinguish envelope from domain and header from domain as they are used differently in SPF, DKIM, DMARC?

**Review by nguyenkims — COMMENTED** [review]
The PR looks really good! I just left some rather nits comments.

File: app/models.py (hunk 1 above)
Comment (acasajus) [reply app/models.py:286]: We're not using it yet. I'll remove it to make it cleaner for now, but as you say the idea was to add all the relevent info for anti-abuse in this object so we decide what to do based on it. But for now I've just added spf status.

File: email_handler.py [hunk 2]
```diff
@@ -2597,18 +2592,29 @@ def _handle(self, envelope: Envelope, msg: Message):
         )
 
         with create_light_app().app_context():
-            ret = handle(envelope, msg)
+            return_status = handle(envelope, msg)
             elapsed = time.time() - start
+            if return_status[0] == "5":
```
[the remark this answers is not in the document]
Comment (acasajus) [reply email_handler.py:2597]: :+1:

File: email_handler.py [hunk 3]
```diff
@@ -541,25 +542,28 @@ def handle_email_sent_to_ourself(alias, from_addr: str, msg: Message, user):
...
-    else:
-        newrelic.agent.record_custom_event("DmarcCheck", {"result": "unknown"})
-
-    if not DMARC_CHECK_ENABLED or not dmarc_result:
+    spam_result = get_spamd_result(msg)
+    if not DMARC_CHECK_ENABLED or not spam_result:
         return None
 
-    if dmarc_result in (
+    from_header = get_header_unicode(msg[headers.FROM])
```
[the remark this answers is not in the document]
Comment (acasajus) [reply email_handler.py:551]: Where is it done? The `from_header` variable is being created in line `handle_forward` and it's not modified afaik. I thought that the from replacement happened in the `forward_email_to_mailbox`

File: app/email_utils.py [hunk 4]
```diff
@@ -1441,19 +1443,26 @@ def save_email_for_debugging(msg: Message, file_name_prefix=None) -> str:
     return ""
 
 
-def get_dmarc_status(msg: Message) -> Optional[DmarcCheckResult]:
-    spam_result = msg.get_all(headers.SPAMD_RESULT)
-    if not spam_result:
+def get_spamd_result(msg: Message) -> Optional[SpamdResult]:
+    spam_result_header = msg.get_all(headers.SPAMD_RESULT)
+    if not spam_result_header:
+        newrelic.agent.record_custom_event("SpamdCheck", {"header": "missing"})
```
[the remark this answers is not in the document]
Comment (acasajus) [reply app/email_utils.py:1449]: This way we can add more information to the event if needed.

File: email_handler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nguyenkims) [reply email_handler.py:551]: Oh you're right, I guess I added this when trying to putting apply_dmarc_policy into forward_email_to_mailbox but it isn't needed now.

**Review by nguyenkims — APPROVED** [review]
LGTM!

**Pull request merged**